src/graphql/fanout.ts

```diff
@@ -0,0 +1,307 @@
+  args: {
+    encryptedMessage: nonNull(
+      arg({
+        type: 'EncryptedMessage',
+      }),
+    ),
+    name: nonNull(
+      arg({
+        type: 'String',
+      }),
+    ),
+    members: nonNull(
+      arg({
+        type: list('FanoutMember'),
+      }),
+    ),
+    splTokenAddresses: arg({
+      type: list('String'),
+    }),
+  },
+  async resolve (_, args, ctx: YogaInitialContext) {
+    let authorityWallet: Wallet = null!;
+    const connection = new Connection(process.env.RPC_ENDPOINT, 'confirmed');
+    let fanoutSdk: FanoutClient;
+    let splFanoutResult: {splTokenAddress: PublicKey | null, splTokenWallet: PublicKey | null }[] = [];
+
+    const keyPairBytes = JSON.parse(
+      decryptEncodedPayload(args.encryptedMessage),
+    ) as number[];
+
+    // Get create wallet from the client secrets
+    try {
+      authorityWallet = new Wallet(
+        Keypair.fromSecretKey(Uint8Array.from(keyPairBytes)),
+      );
+    } catch (e) {
+      return {
+        message: `Error creating wallet from client secret: ${e.message}`,
+      };
+    }
+
+    // Make sure sum of shares is 100
+    if (
+      args.members.reduce((partialSum, a) => partialSum + a.shares, 0) !== 100
+    ) {
+      return {
+        message: 'Total share must be 100',
+      };
+    }
+
+    // Make sure all members are valid
+    if (
+      args.members.filter((m) => new PublicKey(m.publicKey)).length !==
+      args.members.length
+    ) {
+      return {
+        message: 'Invalid public key',
+      };
+    }
+
+    fanoutSdk = new FanoutClient(connection, authorityWallet);
+
+    let init: null | {
+      fanout: PublicKey;
+      nativeAccount: PublicKey;
+    } = null;
+    try {
+      init = await fanoutSdk.initializeFanout({
+        totalShares: 100,
+        name: args.name,
+        membershipModel: MembershipModel.Wallet, // TODO: support other membership models
+      });
+    } catch (e) {
+      return {
+        message: `Error initializing fanout: ${e.message}`,
+      };
+    }
+
+    if (args.splTokenAddresses.length > 0) {
+      for (var i = 0; i < args.splTokenAddresses.length; i++) {
+        try {
+          const {
+            fanoutForMint,
+            tokenAccount,
+          } = await fanoutSdk.initializeFanoutForMint({
+            fanout: init.fanout,
+            mint: new PublicKey(args.splTokenAddresses[i]),
+          });
+          splFanoutResult.push({splTokenAddress: new PublicKey(args.splTokenAddresses[i]), splTokenWallet: tokenAccount})
+        } catch (e) {
+          return {
+            message: `Error initializing fanout for mint: ${e.message}`,
+          };
+        }
+      }
+    }
+
+    // Add members
+    args.members.map(async (member) => {
+      try {
+        await fanoutSdk.addMemberWallet({
+          fanout: init.fanout,
+          fanoutNativeAccount: init.nativeAccount,
+          membershipKey: new PublicKey(member.publicKey),
+          shares: member.shares,
+        });
```

Comment:
It doesnt look like it is idempotent

add member program function - https://github.com/GlassEaters/hydra/blob/main/programs/hydra/src/processors/add_member/wallet.rs#L36

will drop the members into a set for uniqueness 👍🏽

Reply:
Okay array is unique, please review and 👍🏽 if you think its good to go! @kespinola

Reply:
we chatted in slack. need to be cautious about failures and re-running this action since its submitting so many instructions that may not be able to re-run again. If this mutation fails we may not be able to re-run it.